pkg/instancegroups/instancegroups.go

```diff
@@ -308,141 +334,288 @@ func (n *CloudInstanceGroup) RollingUpdate(rollingUpdateData *RollingUpdateClust
-	if err != nil {
-		return fmt.Errorf("error cordoning node node: %v", err)
+}
+
+// DeleteAWSInstance deletes an EC2 AWS Instance.
+func (c *CloudInstanceGroup) DeleteAWSInstance(u *CloudInstanceGroupInstance, instanceId string, nodeName string, cloud awsup.AWSCloud) error {
+
+	if nodeName != "" {
+		glog.Infof("Stopping instance %q, node %q, in AWS ASG %q.", instanceId, nodeName, c.ASGName)
+	} else {
+		glog.Infof("Stopping instance %q, in AWS ASG %q.", instanceId, c.ASGName)
 	}
 
-	err = options.RunDrain()
-	if err != nil {
-		return fmt.Errorf("error draining node: %v", err)
+	request := &autoscaling.TerminateInstanceInAutoScalingGroupInput{
+		InstanceId:                     u.ASGInstance.InstanceId,
+		ShouldDecrementDesiredCapacity: aws.Bool(false),
 	}
 
-	if rollingUpdateData.DrainInterval > time.Second*0 {
-		glog.V(3).Infof("Waiting for %s for pods to stabilize after draining.", rollingUpdateData.DrainInterval)
-		time.Sleep(rollingUpdateData.DrainInterval)
+	if _, err := cloud.Autoscaling().TerminateInstanceInAutoScalingGroup(request); err != nil {
+		if nodeName != "" {
+			return fmt.Errorf("error deleting instance %q, node %q: %v", instanceId, nodeName, err)
+		}
+		return fmt.Errorf("error deleting instance %q: %v", instanceId, err)
 	}
 
 	return nil
+
 }
 
-func (g *CloudInstanceGroup) Delete(cloud fi.Cloud) error {
-	c := cloud.(awsup.AWSCloud)
+func (c *CloudInstanceGroup) Delete(cloud fi.Cloud) error {
+	cloudInterface := cloud.(awsup.AWSCloud)
 
 	// TODO: Graceful?
 
 	// Delete ASG
 	{
-		asgName := aws.StringValue(g.asg.AutoScalingGroupName)
+		asgName := aws.StringValue(c.asg.AutoScalingGroupName)
 		glog.V(2).Infof("Deleting autoscaling group %q", asgName)
 		request := &autoscaling.DeleteAutoScalingGroupInput{
-			AutoScalingGroupName: g.asg.AutoScalingGroupName,
+			AutoScalingGroupName: c.asg.AutoScalingGroupName,
 			ForceDelete:          aws.Bool(true),
 		}
-		_, err := c.Autoscaling().DeleteAutoScalingGroup(request)
+		_, err := cloudInterface.Autoscaling().DeleteAutoScalingGroup(request)
 		if err != nil {
 			return fmt.Errorf("error deleting autoscaling group %q: %v", asgName, err)
 		}
 	}
 
 	// Delete LaunchConfig
 	{
-		lcName := aws.StringValue(g.asg.LaunchConfigurationName)
+		lcName := aws.StringValue(c.asg.LaunchConfigurationName)
 		glog.V(2).Infof("Deleting autoscaling launch configuration %q", lcName)
 		request := &autoscaling.DeleteLaunchConfigurationInput{
-			LaunchConfigurationName: g.asg.LaunchConfigurationName,
+			LaunchConfigurationName: c.asg.LaunchConfigurationName,
 		}
-		_, err := c.Autoscaling().DeleteLaunchConfiguration(request)
+		_, err := cloudInterface.Autoscaling().DeleteLaunchConfiguration(request)
 		if err != nil {
 			return fmt.Errorf("error deleting autoscaling launch configuration %q: %v", lcName, err)
 		}
 	}
 
 	return nil
 }
+
+func (c *CloudInstanceGroup) Duplicate(cluster *api.Cluster, clientSet simple.Clientset, suffix string) (*api.InstanceGroup, error) {
```

Comment:
This seems like a very important function, would be nice to have unit tests for it :)